### PR TITLE
Added MecanumDriveOdometry Python wrapper

### PR DESCRIPTION
**Pull request by ahcorde**
# 🎉 New feature

Closes https://github.com/gazebosim/gz-math/issues/548

## Summary

Added MecanumDriveOdometry

## Checklist
- [x] Signed all commits for DCO
- [x] Added tests
- [ ] Added example and/or tutorial
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [x] Consider updating Python bindings (if the library has them)
- [x] `codecheck` passed (See [contributing](https://gazebosim.org/docs/all/contributing#contributing-code))
- [x] All tests passed (See [test coverage](https://gazebosim.org/docs/all/contributing#test-coverage))
- [x] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Agazebosim+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge** and edit the commit message to match the pull request summary while retaining `Signed-off-by` messages.

### CODE REVIEW

**Comment by azeey** [comment]
Is it okay if we wait till after Harmonic for this?

**Comment by ahcorde** [comment]
this is targeting Fortress, but we can wait no problem

**Comment by azeey** [comment]
> this is targeting Fortress, but we can wait no problem

I understand, but I think we should focus our review efforts on `beta` labeled PRs.

**Comment by ahcorde** [comment]
friendly ping @azeey @scpeters or @adityapande-1995

File: src/python_pybind11/src/MecanumDriveOdometry.cc [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+{
+namespace python
+{
+void defineMathMecanumDriveOdometry(py::module &m, const std::string &typestr)
+{
+  using Class = gz::math::MecanumDriveOdometry;
+  std::string pyclass_name = typestr;
+  py::class_<Class>(m,
+                    pyclass_name.c_str(),
+                    py::buffer_protocol(),
```
Comment (azeey) [inline src/python_pybind11/src/MecanumDriveOdometry.cc:36]: Do we need `py::buffer_protocol` here? If so, doesn't it require implementing `def_buffer` (https://pybind11.readthedocs.io/en/stable/advanced/pycpp/numpy.html)

File: src/python_pybind11/src/MecanumDriveOdometry.cc [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+                    py::dynamic_attr())
```
Comment (azeey) [inline src/python_pybind11/src/MecanumDriveOdometry.cc:37]: Why do we need `py::dynamic_attr`?

File: src/python_pybind11/test/MecanumDriveOdometry_TEST.py [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+        wheelRadius = 0.5
+        wheelCircumference = 2 * math.pi * wheelRadius
+
+        # This is the linear distance traveled per degree of wheel rotation.
+        distPerDegree = wheelCircumference / 360.0
+
+        # Setup the wheel parameters, and initialize
+        odom.set_wheel_params(wheelSeparation, wheelRadius, wheelRadius,wheelRadius)
+        startTime = datetime.datetime.now()
+        odom.init(datetime.timedelta())
```
Comment (azeey) [inline src/python_pybind11/test/MecanumDriveOdometry_TEST.py:42]: Should this use `startTime`?

File: src/python_pybind11/test/MecanumDriveOdometry_TEST.py [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+
+        # Sleep for a little while, then update the odometry with the new wheel
+        # position.
+        time1 = startTime + datetime.timedelta(milliseconds=100)
+        odom.update(Angle(1.0 * math.pi / 180),
+                    Angle(1.0 * math.pi / 180),
+                    Angle(1.0 * math.pi / 180),
+                    Angle(1.0 * math.pi / 180),
+                    time1 - startTime)
```
Comment (azeey) [inline src/python_pybind11/test/MecanumDriveOdometry_TEST.py:51]: The C++ version uses just `time1` here.

File: src/python_pybind11/test/MecanumDriveOdometry_TEST.py [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+        # straight line.
+        self.assertAlmostEqual(0.0, odom.angular_velocity().radian(), delta=1e-3)
+
+        # Sleep again, then update the odometry with the new wheel position.
+        time2 = time1 + datetime.timedelta(milliseconds=100)
+        odom.update(Angle(2.0 * math.pi / 180),
+                    Angle(2.0 * math.pi / 180),
+                    Angle(2.0 * math.pi / 180),
+                    Angle(2.0 * math.pi / 180),
+                    time2 - startTime)
```
Comment (azeey) [inline src/python_pybind11/test/MecanumDriveOdometry_TEST.py:67]: This is also different from the C++ test

File: src/python_pybind11/test/MecanumDriveOdometry_TEST.py [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+
+        # Setup the wheel parameters, and initialize
+        odom.set_wheel_params(wheelSeparation, wheelRadius, wheelRadius,wheelRadius)
+        startTime = datetime.datetime.now()
+        odom.init(datetime.timedelta())
+
+        # Sleep for a little while, then update the odometry with the new wheel
+        # position.
+        time1 = startTime + datetime.timedelta(milliseconds=100)
+        odom.update(Angle(1.0 * math.pi / 180),
```
Comment (azeey) [inline src/python_pybind11/test/MecanumDriveOdometry_TEST.py:47]: Use `math.radians`

**Review by azeey — CHANGES_REQUESTED** [review]

**Review by azeey — CHANGES_REQUESTED** [review]
Looks good overall. Just a few issues with pybind11 options and differences with the C++ test

**Review by azeey — CHANGES_REQUESTED** [review]
Looks good overall. Just a few issues with pybind11 options and differences with the C++ test

**Comment by ahcorde** [comment]
@azeey Time is a little bit different in Python, you can take a look the `DiffDriveOdometry_TEST.py` test,

**Comment by ahcorde** [comment]
friendly ping @azeey

**Comment by azeey** [comment]
> @azeey Time is a little bit different in Python, you can take a look the `DiffDriveOdometry_TEST.py` test,

So from the [pybind11 docs](https://pybind11.readthedocs.io/en/stable/advanced/cast/chrono.html#provided-conversions), it looks like `std::chrono::steady_clock::time_point` is converted to `datetime.timedelta`. This makes it really inconvenient to use since `datetime.timedelta()` always returns a "0" time delta as opposed to the current monotonic (in C++ `steady_clock`) time. One option is to use `datetime.timedelta(time.monotonic())` to initialize `startTime` instead of `datetime.datetime.now()`.

File: src/python_pybind11/src/MecanumDriveOdometry.cc (hunk 1 above)
Comment (azeey) [reply src/python_pybind11/src/MecanumDriveOdometry.cc]: I don't think this has been addressed

**Comment by j-rivero** [comment]
@osrf-jenkins run tests

**Comment by ahcorde** [comment]
@azeey I added your feedback

**Comment by azeey** [comment]
> @azeey I added your feedback

Thanks. I think now the test can be updated to match the C++ version. Specifically, `odom.init(startTime)` instead of `odom.init(datetime.timedelta())` and remove `startTime` from `odom.update` calls.

**Comment by ahcorde** [comment]
> Thanks. I think now the test can be updated to match the C++ version. Specifically, `odom.init(startTime)` instead of `odom.init(datetime.timedelta())` and remove `startTime` from `odom.update` calls.

@azeey Included feedback

File: src/python_pybind11/test/DiffDriveOdometry_TEST.py [hunk 7]
```diff
@@ -38,7 +39,7 @@ def test_diff_drive_odometry(self):
 
         # Setup the wheel parameters, and initialize
         odom.set_wheel_params(wheelSeparation, wheelRadius, wheelRadius)
-        startTime = datetime.datetime.now()
+        startTime = datetime.timedelta(time.monotonic())
         odom.init(datetime.timedelta())
```
Comment (azeey) [inline src/python_pybind11/test/DiffDriveOdometry_TEST.py:43]:
```suggestion
        startTime = datetime.timedelta(time.monotonic())
        odom.init(startTime)
```

then update
`odom.update(..., ..., time1)`

File: src/python_pybind11/test/DiffDriveOdometry_TEST.py [hunk 8]
```diff
@@ -71,7 +72,7 @@ def test_diff_drive_odometry(self):
         self.assertAlmostEqual(0.0, odom.angular_velocity().radian(), delta=1e-3)
 
         # Initialize again, and odom values should be reset.
-        startTime = datetime.datetime.now()
+        startTime = datetime.timedelta(time.monotonic())
         odom.init(datetime.timedelta())
```
Comment (azeey) [inline src/python_pybind11/test/DiffDriveOdometry_TEST.py:76]:
```suggestion
        startTime = datetime.timedelta(time.monotonic())
        odom.init(startTime)
```


then update
`odom.update(..., ..., time1)`

File: src/python_pybind11/test/DiffDriveOdometry_TEST.py
Comment (azeey) [inline src/python_pybind11/test/DiffDriveOdometry_TEST.py:1]: We need to make the `odom.init` and `odom.update` changes to `time2` as well.

**Review by azeey — APPROVED** [review]

**Pull request merged**